src/components/cylc/workflow/Lumino.vue

```diff
@@ -15,187 +15,243 @@ You should have received a copy of the GNU General Public License
-      this.box.update()
-    })
-
-    this.$nextTick(() => {
-      // Attach box panel to DOM:
-      Widget.attach(this.box, this.$refs.main)
-      // Watch for resize of the main element to trigger relayout:
-      resizeObserver.observe(this.$refs.main)
-    })
+  allViews: {
+    type: Map,
+    required: true
   },
+})
 
-  computed: {
-    /**
-     * We want to watch this.views; however, the (newVal, oldVal) args
-     * do not differ when a deeply watched object's properties change.
-     * So here is a workaround.
-     */
-    _views () {
-      return Object.assign({}, this.views)
-    }
-  },
+const emit = defineEmits([
+  'emptied'
+])
+
+/**
+ * Template ref
+ * @type {import('vue').Ref<HTMLElement>}
+ */
+const mainDiv = ref(null)
+
+/**
+ * Mapping of widget ID to the name of view component and its initialOptions prop.
+ *
+ * @type {import('vue').Ref<Map<string, AddViewEvent>>}
+ */
+const views = ref(new Map())
+
+const defaultView = useDefaultView()
+
+// create a box panel, which holds the dock panel, and controls its layout
+const boxPanel = new BoxPanel({ direction: 'left-to-right', spacing: 0 })
+// create dock panel, which holds the widgets
+const dockPanel = new DockPanel()
+boxPanel.addWidget(dockPanel)
+BoxPanel.setStretch(dockPanel, 1)
+
+const resizeObserver = new ResizeObserver(() => {
+  boxPanel.update()
+})
+
+onMounted(() => {
+  // Attach box panel to DOM:
+  Widget.attach(boxPanel, mainDiv.value)
+  // Watch for resize of the main element to trigger relayout:
+  resizeObserver.observe(mainDiv.value)
+  $eventBus.on('add-view', addView)
+  getLayout(props.workflowName)
+})
+
+onBeforeUnmount(() => {
+  resizeObserver.disconnect()
+  $eventBus.off('add-view', addView)
+  saveLayout()
+  // Register with Lumino that the dock panel is no longer used,
+  // otherwise uncaught errors can occur when restoring layout
+  dockPanel.dispose()
+})
 
-  watch: {
-    _views: {
-      deep: true,
-      handler: 'syncWidgets'
+/**
+ * Create a widget and add it to the dock.
+ *
+ * @param {AddViewEvent} view
+ * @param {boolean} onTop
+ */
+const addView = (view, onTop = true) => {
+  const id = uniqueId('widget')
+  const luminoWidget = new LuminoWidget(id, startCase(view.name), /* closable */ true)
+  dockPanel.addWidget(luminoWidget, { mode: 'tab-after' })
+  // give time for Lumino's widget DOM element to be created
+  nextTick(() => {
+    views.value.set(id, view)
+    addWidgetEventListeners(id)
+    if (onTop) {
+      dockPanel.selectWidget(luminoWidget)
     }
-  },
+  })
+}
 
-  methods: {
-    /**
-     * Look for newly added views, creating a corresponding Lumino Widget
-     * for each.
-     */
-    syncWidgets (newVal, oldVal) {
-      for (const [id, item] of Object.entries(newVal)) {
-        if (!(id in oldVal)) {
-          this.addWidget(id, item.view)
-        }
+/**
+ * Remove all the widgets present in the DockPanel.
+ */
+const closeAllViews = () => {
```

Comment:
IMO:

```js
function foo(bar) { ... }
```

Is cleaner than

```js
const foo = (bar) => { ... }
```

What's the motive for the `const` approach? Does it do something different?

Reply:
No preference from me

Reply:
So the only functional difference is the implicit `this` variable?

Reply:
Well, also declarations (`function foo () {}`) are hoisted but expressions (`const foo = () => {}`) are not.